engine/main.ml

```
@@ -224,39 +230,337 @@ let rec parse_lambda lsp =
+      | Isout i -> bprintf buf "Isout %d" i
+      | Isin i -> bprintf buf "Isin %d" i
+    in
+    bprintf buf "{ var=%a; op=%a; }" bprint_accessor pi.var print_op pi.op
+  in
+  let rec bprint_list ~sep bprint buf = function
+    | [] -> ()
+    | [x] -> bprint buf x
+    | x :: xs ->
+       bprintf buf "%a%t%a"
+         bprint x
+         sep
+         (bprint_list ~sep bprint) xs in
+  let break ntabs buf =
+    bprintf buf "\n%s" (BatList.init ntabs (fun _ -> "\t") |> String.concat "") in
+  let rec bprint_tree ntabs buf tree =
+    let sep = break ntabs in
+    match tree with
+    | Leaf (pilist, target_blackbox) ->
+       bprintf buf
+         "Leaf=%s\
+              %t%a"
+         target_blackbox
+         (break ntabs) (bprint_list ~sep bprint_pi) pilist
+    | Node (cst_list) ->
+       bprintf buf
+         "Node=\
+          %t%a"
+         (break ntabs)
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+  and
+  bprint_env ntabs buf env =
+    let bprint_binding buf (key, entry) =
+      let bprint_value buf = function
+        | Accessor a -> bprint_accessor buf a
+        | Addition (v, i) ->
+           bprintf buf "Addition=%s,%i" v i
+        | Function (v, f_tree) ->
+           bprintf buf  "Function=%s,ConstraintTree: %a"
+             v
+             (bprint_tree (ntabs + 1)) f_tree
+        | Catch (e, vars, c_tree) ->
+          bprintf buf "Catch=%d %s,ConstraintTree: %a"
+            e
+            (String.concat " " vars)
+             (bprint_tree (ntabs + 1)) c_tree
+      in
+      bprintf buf "%s: %a" key bprint_value entry
+    in
+    bprint_list ~sep:(break ntabs) bprint_binding buf (SMap.bindings env)
+  in
+  let print_env env =
+    let buf = Buffer.create 42 in
+    bprint_env 0 buf env;
+    BatIO.write_line BatIO.stdout (Buffer.contents buf)
+  in
+  let expand_env variable entry : environment SMap.t = 
+    let _ = match SMap.find_opt variable env with
+      | Some _ -> assert false
+      | None -> ()
```

In these cases you can use `SMap.find` which raises an exception instead of returning an option as `find_opt` does. Your code is not especially nicer.

---

Scrap that: actually you mean `assert (not (SMap.mem variable env));`